bin/mage

```diff
@@ -1,55 +1,268 @@
+  esac
+
+  case "$(uname -m)" in
+    x86_64|amd64) arch="amd64" ;;
+    arm64|aarch64) arch="aarch64" ;;
+    *) log_error "Unsupported architecture: $(uname -m)"; exit 1 ;;
+  esac
+
+  echo "${os}-${arch}"
+}
+
+# Download babashka if not found
+install_babashka() {
+  local platform tmp_dir download_url min_bb_version
+
+  # Get required Babashka version
+  min_bb_version=$(get_min_bb_version)
+
+  dbg_info "Babashka not found at ${BB_EXEC}. Installing version ${min_bb_version}..."
+
+  platform=$(detect_platform)
+  log_info "Detected platform: ${platform}"
+
+  tmp_dir=$(mktemp -d)
+  dbg_info "Created temporary directory: ${tmp_dir}"
+
+  download_url="https://github.com/babashka/babashka/releases/download/v${min_bb_version}/babashka-${min_bb_version}-${platform}.tar.gz"
+  log_info "Download URL: ${download_url}"
+
+  # Create bin directory relative to the script location
+  dbg_info "Creating bin directory: ${BB_BIN_DIR}"
+  mkdir -p "${BB_BIN_DIR}"
+
+  # Download the file first
+  curl -L "${download_url}" -o "${tmp_dir}/babashka.tar.gz"
+
+  if [ ! -f "${tmp_dir}/babashka.tar.gz" ]; then
+    log_error "Download failed! Could not find downloaded file"
+    exit 1
+  fi
+
+  log_info "Download complete. Extracting archive..."
+  tar -xzf "${tmp_dir}/babashka.tar.gz" -C "${tmp_dir}"
+
+  # List the temp directory contents to see if extraction worked
+  dbg_info "Files in temp directory after extraction:"
+  ls -la "${tmp_dir}"
+
+  if [ -f "${tmp_dir}/bb" ]; then
+    dbg_info "Moving bb to ${BB_EXEC}"
+    mv "${tmp_dir}/bb" "${BB_EXEC}"
+    chmod +x "${BB_EXEC}"
+
+    # Verify installation
+    if [ -x "${BB_EXEC}" ]; then
+      dbg_info "Testing installed binary..."
+      "${BB_EXEC}" --version
+      log_success "Babashka ${min_bb_version} successfully installed at ${BB_EXEC}"
+    else
+      log_error "Installation failed! ${BB_EXEC} is not executable"
+      exit 1
+    fi
+  else
+    log_error "Extraction failed! Could not find bb executable in extracted files - are you sure the version is correct?"
+    exit 1
+  fi
+
+  rm -rf "${tmp_dir}"
+}
+
+# Check if babashka is available at ./bin/bb
+check_babashka() {
+  dbg_info "Checking for Babashka at ${BB_EXEC}"
+
+  if [ -x "${BB_EXEC}" ]; then
+    dbg_info "Found executable at ${BB_EXEC}"
+    # Check if installed version meets minimum requirements
+    local installed_version min_bb_version
+    min_bb_version=$(get_min_bb_version)
+
+    installed_version=$("${BB_EXEC}" --version 2>/dev/null | head -n 1 | grep -o '[0-9]\+\.[0-9]\+\.[0-9]\+')
+
+    if [ -z "${installed_version}" ]; then
+      log_warning "Could not determine installed Babashka version"
+      return 1
+    fi
+
+    dbg_info "Installed version: ${installed_version}, Required: ${min_bb_version}"
+
+    # Proper semantic version comparison
+    local comparison_result
+    comparison_result=$(compare_versions "${installed_version}" "${min_bb_version}")
+
+    if [ "$comparison_result" -lt "0" ]; then
```

Comment:
uhhh weird but how about using `compare_version` you just move comparison directly here and we'll be able to forget about weird `echo` and this damned `-lt` 😁